src/ring_theory/ideal/nc_jacobson_radical.lean

```diff
@@ -0,0 +1,250 @@
+let ⟨a, _⟩ := h in ⟨units.op_equiv.symm (op a), by simpa⟩
+
+lemma is_unit.unop {x : Mᵐᵒᵖ} (h : is_unit x) : is_unit (unop x) :=
+let ⟨a, _⟩ := h in ⟨unop a.op_equiv, by simpa⟩
+
+lemma is_unit_op_iff_is_unit {a : M} : is_unit (op a) ↔ is_unit a :=
+⟨λ h, unop_op a ▸ h.unop, is_unit.op⟩
+
+lemma is_unit_unop_iff_is_unit {x : Mᵐᵒᵖ} : is_unit (unop x) ↔ is_unit x :=
+⟨λ h, op_unop x ▸ h.op, is_unit.unop⟩
+
+lemma is_unit_iff_has_two_sided_inv :
+  is_unit a ↔ ∃ x : M, a * x = 1 ∧ x * a = 1 :=
+⟨λ ⟨⟨a, x, hax, hxa⟩, rfl⟩, ⟨x, hax, hxa⟩, λ ⟨x, hax, hxa⟩, ⟨⟨a, x, hax, hxa⟩, rfl⟩⟩
+
+/-- An element `a : M` of a monoid has a left inverse
+if there is some `x : M` satisfying `x * a = 1`. -/
+def has_left_inv (a : M) : Prop := ∃ x : M, x * a = 1
+
+/-- An element `a : M` of a monoid has a right inverse
+if there is some `x : M` satisfying `a * x = 1`. -/
+def has_right_inv (a : M) : Prop := ∃ x : M, a * x = 1
+
+/-- An element of a monoid is a unit
+if and only if it has both a left inverse and a right inverse. -/
+theorem is_unit_iff_has_left_inv_right_inv :
+  is_unit a ↔ has_left_inv a ∧ has_right_inv a :=
+⟨λ h, ⟨h.exists_left_inv, h.exists_right_inv⟩,
+  λ ⟨⟨x, hxa⟩, ⟨y, hay⟩⟩, ⟨⟨a, x, by { convert hay,
+  rw [←mul_one x, ←hay, ←mul_assoc, hxa, one_mul] }, hxa⟩, rfl⟩⟩
+
+/-- An element of a monoid is a unit
+if it has a left inverse which also has a left inverse. -/
+theorem is_unit_of_has_left_inv_of_has_left_inv
+  (h₁ : b * a = 1) (h₂ : has_left_inv b) : is_unit a :=
+let ⟨c, hcb⟩ := h₂ in ⟨⟨a, b, by { convert hcb,
+  rw [←one_mul a, ←hcb, mul_assoc, h₁, mul_one] }, h₁⟩, rfl⟩
+
+end monoid
+
+section ring
+variables {R : Type u} {a : R} {b : R} [ring R]
+/-! ### Some lemmas on inverses in rings -/
+
+lemma has_left_inv.one_add_mul_swap
+  (h : has_left_inv (1 + a * b)) : has_left_inv (1 + b * a) :=
+begin
+  obtain ⟨u, hu⟩ := h,
+  existsi 1 - b * u * a,
+  calc (1 - b * u * a) * (1 + b * a)
+        = 1 + b * a - b * (u * (1 + a * b)) * a : by noncomm_ring
+    ... = 1 : by rw [hu, mul_one, add_sub_cancel],
+end
+
+lemma has_right_inv.one_add_mul_swap
+  (h : has_right_inv (1 + a * b)) : has_right_inv (1 + b * a) :=
+begin
+  obtain ⟨u, hu⟩ := h,
+  existsi 1 - b * u * a,
+  calc (1 + b * a) * (1 - b * u * a)
+        = 1 + b * a - b * ((1 + a * b ) * u) * a : by noncomm_ring
+    ... = 1 : by rw [hu, mul_one, add_sub_cancel],
+end
+
+lemma is_unit.one_add_mul_swap
+  (h : is_unit (1 + a * b)) : is_unit (1 + b * a) :=
+begin
+  rw is_unit_iff_has_left_inv_right_inv at *,
+  obtain ⟨h₁, h₂⟩ := h,
+  exact ⟨h₁.one_add_mul_swap, h₂.one_add_mul_swap⟩,
+end
+
+end ring
+
+namespace ideal
+section nc_jacobson_radical
+variables {R : Type u} [ring R]
+/-! ### Jacobson radical of a ring -/
+
+/--
+For a ring `R`, `nc_jacobson R` is the Jacobson radical of `R`, that is,
+the intersection of all maximal left ideals of of `R`. Note that we use left ideals.
+-/
+def nc_jacobson (R : Type u) [ring R] : ideal R := jacobson ⊥
+
```

Comment:
Looking at the definition in the commutative case, I think it makes sense to make this a separate definition in the non-commutative case. This seems to be a confusion that has appeared elsewhere, see e.g. [https://math.stackexchange.com/questions/1493997/the-jacobson-radical-of-an-ideal](https://math.stackexchange.com/questions/1493997/the-jacobson-radical-of-an-ideal). One could follow up defining the Jacobson radical of a left module.

Reply:
I wonder if `ring_jacobson` would be a better name? No particular need to emphasise noncommutativity in the name. Simply not having a name and writing `jacobson ⊥` still seems viable to me as well.

But this is fine as is: I don't want to bikeshed!!

Reply:
Personally, I like `ring_jacobson` instead of `nc_jacobson`. Another possibility, I communicated already to Johan, is to rename the (previously commutative) `jacobson` to `ideal.jacobson` or something of this kind.

Reply:
I like the idea of having `ring.jacobson` being the global version and `ideal.jacobson` being the ideal-specific version; sort of like how we have `module.finite` and `submodule.fg`.

Reply:
I agree with using `ring.jacobson R` for the Jacobson radial of a ring `R` and `ideal.jacobson I` for `I : ideal R` for the intersection of maximal (left) ideals containing `I`.
And in this way I guess no confusion occurs (thanks to type check!).

I think the name for the latter is already `ideal.jacobson` (it's in namespace `ideal`).
Also I think that working in namespace `ideal` for this PR's Jacobson radical of a ring is not appropriate. For example, obviously I don't want `I.nc_jacobson` for `I : ideal R` (which yields a possible confusion with `I.ideal`), and we only consider one fixed ideal, Jacobson radical, so statements in this file are ring-theoretic ones, not ideal-theoretic ones (except for one lemma `one_add_mul_self_mem_maximal_of_not_mem_maximal`).

By these reason,
* I change working namespace from `ideal` to `ring` (except for one lemma, which is still in namespace `ideal`).
* I rename previous `ideal.nc_jacobson` to `ring.jacobson`.